Катцин Валерий

valerin5 opened the pull request.

Commit:
commit

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

Commit:
fix

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

Commit:
fix lint and clean

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

SemyonMakhaev commented on lib.js:
```diff
@@ -1,15 +1,88 @@
+        bestFriends: [],
+        otherFriends: []
+    };
+    friends.forEach(friend => {
```
Можно сформировать объект через `reduce`

SemyonMakhaev commented on lib.js:
```diff
@@ -1,15 +1,88 @@
+};
+
+Iterator.prototype.next = function () {
+    return (this.done()) ? null : this.friends[this.place++];
```
Можно без скобочек
```suggestion
    return this.done() ? null : this.friends[this.place++];
```

SemyonMakhaev commented on lib.js:
```diff
@@ -19,15 +92,35 @@ function Iterator(friends, filter) {
+        result = result.concat(halfResult);
+    }
+    this.friends = result;
+    this.place = 0;
```
Кажется, что `place` здесь можно не задавать, потому что он определяется в родительском классе `Iterator` в строке [74](https://github.com/urfu-2018/javascript-task-5/pull/53/files#diff-a0718ce45a590dd715d1a7ed856d3ebbR74)

valerin5 replied:
Не получается его убрать, скорее всего из-за области видимости

valerin5 replied:
Да и мы конструктор переопределяем

SemyonMakhaev replied:
Не очень понял про область видимости. Здесь ведь поле наследуется от прототипа, область видимости - это про доступ к переменным, объявленным внутри функции, а `place` - это собственное свойство функции `Iterator`.

Рекомендую: https://learn.javascript.ru/closures

SemyonMakhaev reviewed (changes requested):
🍅

Commit:
fix

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

Commit:
test

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

Commit:
test 2

honest-hrundel commented:
:tomato: Пройдено тестов 18 из 24

Commit:
test3

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

Commit:
test4

honest-hrundel commented:
:tomato: Пройдено тестов 18 из 24

Commit:
test5

honest-hrundel commented:
:tomato: Пройдено тестов 18 из 24

Commit:
пример теста

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

SemyonMakhaev commented on lib.js:
```diff
@@ -1,15 +1,96 @@
+    while (objectFriends.otherFriends.length !== 0) {
+        const length = objectFriends.otherFriends.length;
+        let friendsFriend = [];
+        objectFriends.bestFriends.forEach(friend => {
```
Получится ли сформировать массив `friendsFriend` через `reduce`?

SemyonMakhaev commented on lib.js:
```diff
@@ -1,15 +1,96 @@
+    }
+    result = result.map((arr, index) => {
+        arr = arr.map(elem => {
+            elem.level = index + 1;
```
Можно не вычислять `index + 1` на каждой итерации, если положить значение в переменную перед циклом

SemyonMakhaev commented on lib.js:
```diff
@@ -1,15 +1,96 @@
+
+        return arr;
+    });
+    // console.info(result);
```
👀

SemyonMakhaev commented on lib.spec.js:
```diff
@@ -78,6 +78,27 @@ describe('Итераторы', () => {
         ]);
     });
 
+    it('repeat чуть-чуть по другому', () => {
```
Молодец, что разобрался! 👍

SemyonMakhaev reviewed (changes requested):
🍅

Commit:
fix2

Пришлось выключить линт в одном месте, ибо после исправлений самого места не касающегося он начал ругаться

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

SemyonMakhaev reviewed (approved):
🍏
